fix(csi-driver): support getting device size from devpath

niladrih opened the pull request:
Changes:
- the Device::lookup() API doesn't return a dev-name, but instead returns a dev-path. Changed the size function to accept either.
- the `failure!` macro prints errors event if they are never encountered, when used with the `ok_or()` combinator. Using the `ok_or_else()` combinator instead.
- Adjusting the expected capacity of the nexus during NodeExpandVolume call to be at most 10MiB less than the requested capacity.

niladrih committed:
fix(csi-node): support getting device size from devpath

Signed-off-by: Niladri Halder <niladri.halder26@gmail.com>

tiagolobocastro commented on control-plane/csi-driver/src/bin/node/dev.rs:
```diff
@@ -142,11 +142,21 @@ impl Device {
 /// Get the block device capacity size from the sysfs block count.
-pub(crate) fn get_size_from_dev_name<N: AsRef<Path>>(dev_name: N) -> Result<usize, DeviceError> {
+/// Arg can be /dev/device-name style dev-path or just the device name itself.
+pub(crate) fn get_sysfs_size_for_device<N: AsRef<Path>>(device: N) -> Result<usize, DeviceError> {
```
```suggestion
pub(crate) fn sysfs_dev_size<N: AsRef<Path>>(device: N) -> Result<usize, DeviceError> {
```

tiagolobocastro commented on control-plane/csi-driver/src/bin/node/node.rs:
```diff
@@ -512,21 +510,19 @@ impl node_server::Node for Node {
-                "failed to find a device for volume {}",
-                vol_uuid
-            ))?
+            .ok_or(Status::invalid_argument(format!(
```
ok_or_else would fix this too

niladrih committed:
fix(csi-node): use ok_or_else instead of ok_or with failure! macro

Signed-off-by: Niladri Halder <niladri.halder26@gmail.com>

niladrih committed:
feat(csi-node): trim 10MiB off the required_bytes
When processing the NodeExpandVolume CSI call, we compare the sysfs size
of the volume nexus against the requested capacity. Shaving 10MiB off the
requested capacity puts the number in the ballpark range of the capacity
we expect from the nexus after the ControllerExpandVolume has expanded it.

Signed-off-by: Niladri Halder <niladri.halder26@gmail.com>

tiagolobocastro commented on control-plane/csi-driver/src/bin/node/node.rs:
```diff
@@ -536,7 +540,7 @@ impl node_server::Node for Node {
         // equal to this corrected capacity. This is required because we are not comparing the
         // required capacity against the REST API Volume resource.
-        const MAX_NEXUS_CAPACITY_DIFFERENCE: i64 = 5 * 1024 * 1024;
+        const MAX_NEXUS_CAPACITY_DIFFERENCE: i64 = 10 * 1024 * 1024;
```
hmm might want to do the opposite and add 10 to device_capacity instead? Otherwise what if resize is to < 10MiB ?

tiagolobocastro replied:
we could also add check on control-plane that resize must be >10MiB difference? Otherwise we could have a whole here where we'll simply not expand?

niladrih replied:
> what if resize is to < 10MiB

It's an `i64`, should be able to work. Moving the 10MiB to the other side anyway.

tiagolobocastro replied:
it would underflow...

dsharma-dc commented on control-plane/csi-driver/src/bin/node/node.rs:
```diff
@@ -536,7 +540,7 @@ impl node_server::Node for Node {
         // down by 5 MiB so that the capacity of the device is verified to be greater than or
         // equal to this corrected capacity. This is required because we are not comparing the
         // required capacity against the REST API Volume resource.
-        const MAX_NEXUS_CAPACITY_DIFFERENCE: i64 = 5 * 1024 * 1024;
```
Comment needs update saying 10MiB includes additional leeway on top of 5MiB.

niladrih committed:
feat(csi-driver): add expansion logs

Signed-off-by: Niladri Halder <niladri.halder26@gmail.com>

tiagolobocastro commented on control-plane/csi-driver/src/bin/controller/controller.rs:
```diff
@@ -981,6 +981,7 @@ impl rpc::csi::controller_server::Controller for CsiControllerSvc {
         request: tonic::Request<ControllerExpandVolumeRequest>,
     ) -> Result<tonic::Response<ControllerExpandVolumeResponse>, tonic::Status> {
         let request = request.into_inner();
+        info!("Volume expansion request: {:?}", &request);
```
in line with existing methods:
```suggestion
        tracing::trace!(volume.uuid = %args.volume_id, request = ?args);
```

tiagolobocastro commented on control-plane/csi-driver/src/bin/controller/controller.rs:
```diff
@@ -1020,6 +1021,10 @@ impl rpc::csi::controller_server::Controller for CsiControllerSvc {
+        info!(
+            size_bytes = requested_size,
+            "Expansion succeeded for volume {}", &vol_uuid
+        );
```
```suggestion
        debug!(
            size_bytes = requested_size,
            "Expansion succeeded for volume {vol_uuid}"
        );
```

tiagolobocastro commented on control-plane/csi-driver/src/bin/node/node.rs:
```diff
@@ -475,6 +475,7 @@ impl node_server::Node for Node {
         //===============================CsiAccessType=============================================
 
         let request = request.into_inner();
+        info!("Volume expansion request: {:?}", &request);
```
Again, keep with existing ones..
Also might want to move this up?

```suggestion
        trace!("node_expand_volume {request:?}");
```

niladrih replied:
Moved this up, keeping the request, instead of shadowing it.

tiagolobocastro commented on control-plane/csi-driver/src/bin/node/node.rs:
```diff
@@ -590,6 +603,10 @@ impl node_server::Node for Node {
 
+        info!(
+            size_bytes = required_bytes,
+            "Expansion succeeded for volume {}", &vol_uuid
```
```suggestion
            "Expansion succeeded for volume {vol_uuid}"
```

niladrih replied:
Made this a debug.

niladrih commented:
bors merge

tiagolobocastro commented on control-plane/csi-driver/src/bin/node/node.rs:
```diff
@@ -555,20 +561,26 @@ impl node_server::Node for Node {
             // volume_capability AccessType says this is not a 'Mount' type. Nothing to do.
-            CsiAccessType::NotAFilesystem => return success_result,
+            CsiAccessType::NotAFilesystem => {
+                info!(
```
```suggestion
                debug!(
```

niladrih commented:
bors merge
